### ISSUE TITLE
offchain XCMP: Merkle tree of para headers in beefy ignores Coretime chains

### ISSUE THREAD

**Issue by eskimor** [paritytech/polkadot-sdk#4737]
We only collect head data for chains which are registered as legacy parachains [here](https://github.com/paritytech/polkadot-sdk/blob/d2fd53645654d3b8e12cbf735b67b93078d70113/polkadot/runtime/rococo/src/lib.rs#L1304).

This could be easily fixed by iterating the [headers](https://github.com/paritytech/polkadot-sdk/blob/d2fd53645654d3b8e12cbf735b67b93078d70113/polkadot/runtime/rococo/src/lib.rs#L1307) directly. The issue is the number of registered parachains could potentially grow very large. It is economically bounded by currently rather high deposits, but there is already discussions ongoing about reducing it.

We we will build this merkle tree on each and every block so it should stay not only bounded but also rather cheap.

I think the best way to fix this is the following:

1. Make the merkle tree cheaply updatable and store it in state.
2. Don't re-compute the entire tree each block, but instead update it only for parachains actually advancing their head data.

Alternatively ... use state proofs directly and don't bother updating this at all?

**Comment by eskimor** [comment]
@ordian  ... we will need a fix here, otherwise snowbridge will cease to work with chains moving to Coretime  😬

**Comment by eskimor** [comment]
A potential intermediate fix:

Define a constant like this:

```rust
/// This is intermediate "fix" for this issue: https://github.com/orgs/paritytech/projects/119/views/20?pane=issue&itemId=66568838
///
/// It does not actually fix it, but makes the worst case better. Without that limit someone could completely DoS the
/// relay chain by registering a ridiculously high amount of paras. With this limit the same attack could lead to some
/// parachains ceasing to being able to communicate via snowbridge.
const MAX_PARA_HEADERS_IN_PROOF: usize = 512;
```

and use it in benchmarking [here](https://github.com/paritytech/polkadot-sdk/blob/4004742e85f698c2ad84b3e4ab8ea11bbf28135b/bridges/bin/runtime-common/src/parachains_benchmarking.rs#L39) and change the code [here](https://github.com/paritytech/polkadot-sdk/blob/d2fd53645654d3b8e12cbf735b67b93078d70113/polkadot/runtime/rococo/src/lib.rs#L1304) to something like this:

```rust
let mut para_heads = parachains_paras::Heads::<Runtime>::iter().take(MAX_PARA_HEADERS_IN_PROOF).map(|(id, head)| (id.into(), head.0));
//... sort and building of tree.

```

**Comment by acatangiu** [comment]
Cc @vgeddes @alistair-singh

ethereum side parachain proof contracts will likely need adapting depending on the chosen solution

**Comment by coax1d** [comment]
>     1. Make the merkle tree cheaply updatable and store it in state.
> 
>     2. Don't re-compute the entire tree each block, but instead update it only for parachains actually advancing their head data.

@Lederstrumpf and I discussed this exact thing when discussing using this tree in the MMD XCMP design. The best solution is making this thing update-able. Currently the `binary_merkle_tree` crate is honestly extremely simple in its features and the implementation is well.. meh could use some work.

**Comment by vgeddes** [comment]
The workaround in #4751 should work for Snowbridge at least. Have reviewed it. Note that we really only need to verify BridgeHub headers.

If in future we upgrade the "intermediate fix", it needs to be done in a phased manner, to allow us to upgrade our Ethereum contracts without breaking the bridge:
1. Upgrade the relay chain with the "proper fix" while _keeping_ the "intermediate fix"
2. Upgrade the ethereum contracts to use the "proper fix".
3. Remove "intermediate fix" 

cc @seunlanlege not sure how this impacts your bridge.

**Comment by seunlanlege** [comment]
Yeah state proofs are impossible to verify on ethereum. Especially because of the blake2b hash function.

I think it's fine to place a bound on the number of parachain headers. Ideally deregistered parachains should have their head data pruned from the latest storage entry so it doesn't grow too big.

**Comment by ordian** [comment]
> Ideally deregistered parachains should have their head data pruned from the latest storage entry so it doesn't grow too big

this is [already](https://github.com/paritytech/polkadot-sdk/blob/7f7f5fa857502b6e3649081abb6b53c3512bfedb/polkadot/runtime/parachains/src/paras/mod.rs#L1332) the case.
